File: cpp/tests/transport/uart_test_linux/uart_test_linux.cpp
#include <cstdio>
#include <cstring>
#include "UARTTransportLinux.h"

// Assumes a loopback jumper bridging TXD and RXD on the UART port under test.
#ifndef TEST_UART_PORT
#define TEST_UART_PORT "/dev/ttyS0"
#endif
#ifndef TEST_UART_BAUD
#define TEST_UART_BAUD 9600
#endif

static int passed = 0;
static int failed = 0;

static void check_true(const char* label, bool condition) {
    if (condition) { printf("PASS %s\n", label); passed++; }
    else           { printf("FAIL %s\n", label); failed++; }
}

int main() {
    UARTTransportLinux transport(TEST_UART_PORT, TEST_UART_BAUD);

    const uint8_t payload[1] = {0x42};
    transport.write(payload, 1);
    check_true("write accepted", true);

    uint8_t buf[1] = {0};
    transport.read(buf, 1);
    check_true("read returns 1 byte", true);
    check_true("loopback byte matches", buf[0] == 0x42);

    const uint8_t cmd[2] = {0xA5, 0x5A};
    uint8_t resp[2] = {0, 0};
    transport.write_read(cmd, 2, resp, 2);
    check_true("write_read loopback matches", resp[0] == 0xA5 && resp[1] == 0x5A);

    printf("===DONE: %d passed, %d failed===\n", passed, failed);
    return failed == 0 ? 0 : 1;
}

File: python/tests/transport/uart_test_linux.py
import os
import sys

from periph.transport.uart_linux import UARTTransport

UART_PORT     = os.environ.get('UART_PORT',     '/dev/ttyS0')
UART_BAUDRATE = int(os.environ.get('UART_BAUD', '9600'))

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# Assumes a loopback jumper bridging TXD and RXD on the UART port.
transport = UARTTransport(UART_PORT, baudrate=UART_BAUDRATE, timeout_s=1.0)

transport.write(bytes([0x42]))
check_true('write accepted', True)

data = transport.read(1)
check_true('read returns 1 byte', len(data) == 1)
check_true('loopback byte matches', data[0] == 0x42)

result = transport.write_read(bytes([0xA5, 0x5A]), 2)
check_true('write_read returns 2 bytes', len(result) == 2)
check_true('write_read loopback matches', result == bytes([0xA5, 0x5A]))

transport.close()

print('===DONE: {} passed, {} failed==='.format(passed, failed))
sys.exit(0 if failed == 0 else 1)
